### PR TITLE
generate metadata

### PR DESCRIPTION
**Pull request by jackyscript**
#143 

Hi @patmood, refering to the issue #143 here. This is an attempt to defer field length metadata from pocketbase to use for example input validation in the frontend. Basically I wanted to provide the possibility to validate field input data in the frontend using the metadata given in the pocketbase backend.
I am not very confident that this handles every edge case, I think this pull request could benefit from more tests.

When quickly testing this locally I did the following:

- Adjust the `Everything` collection in the `test/pb.schema.json` file locally and set the `max` property to `350` (just for the sake of an example)
- Run `npm run build`
- Run `node dist/index.js --json ./test/pb_schema.json --out ./pocketbase-types.ts --constraints `

The result is a file called `pocketbase-types.constraints.ts` (you can adjust the filename using cli param `--constraintsOut`):

```typescript
/**
* This file was @generated using pocketbase-typegen
*/

export const EverythingFieldConstraints = {
  text_field: { max: 350 }
}

export const UsersFieldConstraints = {
  name: { max: 255 }
}

```

For context: When working in the frontend with validation you could use this generated metadata to setup the `maxLength` property of an input field or some validation logic, that the input does not exceed a certain text length, just as an example:
- https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Attributes/maxlength
- https://vuetifyjs.com/en/components/text-fields/#validation

I hope this is somewhat useful. feel free to leave questions and comments, they are much appreciated! 

EDIT: After the first review and the scope changes resulting from that, I updated the PR's title.

### CODE REVIEW

File: src/lib.ts [hunk 1]
```diff
@@ -109,3 +112,62 @@ export function createResponseType(
...
+      const parts: string[] = []
+      if (typeof field.min === "number" && field.min !== 0) {
+        parts.push(`min: ${field.min}`)
+      }
+      if (typeof field.max === "number" && field.max !== 0) {
+        parts.push(`max: ${field.max}`)
+      }
+
+      if (parts.length === 0) return null
+      return `  ${field.name}: { ${parts.join(", ")} }`
```
Comment (patmood) [inline src/lib.ts]: Make sure this outputs in the same format as the other types, ie fields separated by newline, tab indent etc

File: src/lib.ts [hunk 2]
```diff
@@ -109,3 +112,62 @@ export function createResponseType(
...
+  const fieldsToIgnore = ["id", "password", "tokenKey", "token"]
+  const textBasedTypes = ["text", "email", "url", "number"]
+
+  const constraints = schema
+    .filter((field) => {
+      if (field.system || fieldsToIgnore.includes(field.name)) return false
+      if (!textBasedTypes.includes(field.type)) return false
+      const hasMin = typeof field.min === "number" && field.min !== 0
+      const hasMax = typeof field.max === "number" && field.max !== 0
+      return hasMin || hasMax
```
Comment (patmood) [inline src/lib.ts]: can remove this as it's handled below

File: src/lib.ts [hunk 3]
```diff
@@ -109,3 +112,62 @@ export function createResponseType(
...
+
+  if (constraints.length === 0) return ""
+
+  const typeName = toPascalCase(name)
+  return `export const ${typeName}FieldConstraints = {
+${constraints.join(",\n")}
+}`
+}
+
+export function generateConstraints(results: Array<CollectionRecord>): string {
```
Comment (patmood) [inline src/lib.ts]: Lets match the existing pattern by keeping this function here, but put `createFieldConstraints` in a new file "metadata.ts"

File: src/lib.ts [hunk 4]
```diff
@@ -109,3 +112,62 @@ export function createResponseType(
...
+        if (constraints) fieldConstraints.push(constraints)
+      }
+    })
+
+  const fileParts = [
+    EXPORT_COMMENT,
+    fieldConstraints.length > 0
+      ? fieldConstraints.join("\n")
+      : "// No field constraints found",
+  ]
```
Comment (patmood) [inline src/lib.ts]: FYI I would also like to add the existing enum objects here (select field options, collection names) and deprecate them in the types file, but it can/should be done in a separate PR.

File: src/cli.ts [hunk 5]
```diff
@@ -60,5 +60,14 @@ export async function main(options: Options) {
     sdk: options.sdk ?? true,
   })
   await saveFile(options.out, typeString)
+
+  // Generate and save constraints if requested
+  if (options.constraints) {
+    const constraintsString = generateConstraints(schema)
+    const constraintsFile =
+      options.constraintsOut ?? options.out.replace(".ts", ".constraints.ts")
```
Comment (patmood) [inline src/cli.ts]: set a default value "pocketbase-metadata.ts" in index.ts (same way `--out` options is handled).

**Review by patmood — COMMENTED** [review]
Hey @jackyscript,

Thanks for the PR! Overall looks good, can you please address the comments and add some tests for the new behavior?

File: src/lib.ts (hunk 1 above)
Comment (jackyscript) [reply src/lib.ts]: Adjusted that, thanks!

File: src/lib.ts (hunk 2 above)
Comment (jackyscript) [reply src/lib.ts]: Good catch!

**Comment by patmood** [comment]
Hey @jackyscript, after seeing the latest version with all fields, I realise it's exposing a lot of your schema in JS. I think a better option is to fetch the collection from the API on the client (if it has permission), and use the data from that to validate.

If there's no user permission, you could have a simple script in your project to fetch the collections and filter them by only the collection/fields you want.

How does that sound?

**Comment by jackyscript** [comment]
Hi @patmood thank you for your thoughts, I can see your concerns regarding information exposure.

Given I had a very specific use case in mind, this approach here seems to be a bit too much also for the scope of this project, I will go with your script idea, that way, frontend developers can utilize the metadata for validation right away.

Let us close this pull request then, again thank you for your feedback and review.
